iceoryx_utils/source/posix_wrapper/timer.cpp

```diff
@@ -254,21 +306,26 @@ cxx::expected<TimerError> Timer::OsTimer::start(const RunMode runMode, const Cat
         interval.it_interval.tv_nsec = 0;
     }
 
+    auto& handle = OsTimer::s_callbackHandlePool[m_callbackHandleIndex];
+    // setting m_isTimerActive after timer_settime could lead to false negatives in a check which decides whether
+    // the callback should be executed,
+    // setting it beforehand leads to false positives which can be dealt with and avoid this problem
+    auto wasActive = handle.m_isTimerActive.exchange(true, std::memory_order_relaxed);
+    handle.m_catchUpPolicy = catchUpPolicy;
+
     // Set the timer
     auto result = cxx::makeSmartC(
         timer_settime, cxx::ReturnMode::PRE_DEFINED_ERROR_CODE, {-1}, {}, m_timerId, 0, &interval, nullptr);
 
     if (result.hasErrors())
     {
+        // undo optimistically setting m_isTimerActive before
+        // not entirely safe against concurrent starts, we cannot detect these in this way
+        // needs extensive refactoring to achieve this (e.g. start and stop with mutex protection?)
+        handle.m_isTimerActive.exchange(wasActive, std::memory_order_relaxed);
```

Comment:
Could be a store. But we may miss concurrent timer starts/stops which have set it to another value. Maybe use a compare exchange.

Reply:
I think this was never designed to have concurrent start/stops. If it's done this way, you have the problem you described a while ago with the callback beeing executed before this flag is set and then the callback doesn't do anything because the flag is still false

Reply:
If I remember right, one problem is when a timer is modified (especially restarted) within its callback. At least we should mentioned it somewhere that it might be risky to do that?!

Reply:
Risky is an understatement. It was not race-free to do this before. I think I improved this issue (somewhat), but there will be probably more problems. And as @marthtz says it was used like this internally, hence the fixes.